pytorch_lightning/trainer/trainer.py

```diff
@@ -1223,25 +1226,18 @@ def _log_hyperparams(self) -> None:
             self.logger.log_graph(self.lightning_module)
             self.logger.save()
 
-    def _post_dispatch(self):
-        # these `teardown` calls are here instead of in `_call_teardown_hook` since they are internal teardowns
-        # which need to happen before.
+    def _teardown(self):
+        """This is the Trainer's internal teardown, unrelated to the `teardown` hooks in LightningModule and
+        Callback; those are handled by :meth:`_call_teardown_hook`."""
         self.training_type_plugin.post_dispatch(self)
         self.accelerator.teardown()
         self._data_connector.teardown()
         self._active_loop.teardown()
         self.logger_connector.teardown()
         self.signal_connector.teardown()
 
-    def _dispatch(self) -> Any:
-        if self.evaluating:
-            return self.training_type_plugin.start_evaluating(self)
-        elif self.predicting:
-            return self.training_type_plugin.start_predicting(self)
-        else:
-            return self.training_type_plugin.start_training(self)
-
     def run_stage(self):
+        self.training_type_plugin.barrier("run-stage")
```

Comment:
IMO, I would move this logic within the run function or its own method. More calls might need to be moved in dispatch.

Reply:
+1 can we deprecate run_stage()?

Reply:
created a issue to discuss further. https://github.com/PyTorchLightning/pytorch-lightning/issues/10987

Reply:
Thanks @four4fish, will follow up on it in  #11000